src/libraries/System.Runtime/tests/System.Threading.Tasks.Tests/MethodCoverage.cs

```diff
@@ -352,5 +352,581 @@ public static void FromAsync()
+                Assert.False(twa.IsCompleted);
+                cts.Cancel();
+                Assert.True(twa.IsCompleted);
+                Assert.True(twa.IsCanceled);
+            }
+
+            // Generic, first completes first
+            {
+                var t1 = new TaskCompletionSource<int>();
+                var t2 = new TaskCompletionSource<int>();
+
+                var cts = new CancellationTokenSource();
+                cts.Token.Register(() => t1.TrySetCanceled());
+                cts.Token.Register(() => t2.TrySetCanceled());
+
+                Task<int[]> twa = Task.WhenAll(t1.Task, t2.Task);
+                Assert.False(twa.IsCompleted);
+                cts.Cancel();
+                Assert.True(twa.IsCompleted);
+                Assert.True(twa.IsCanceled);
+            }
+
+            // Non-generic, second completes first
+            {
+                var t1 = new TaskCompletionSource<int>();
+                var t2 = new TaskCompletionSource<int>();
+
+                var cts = new CancellationTokenSource();
+                cts.Token.Register(() => t2.TrySetCanceled());
+                cts.Token.Register(() => t1.TrySetCanceled());
+
+                Task twa = Task.WhenAll((Task)t1.Task, (Task)t2.Task);
+                Assert.False(twa.IsCompleted);
+                cts.Cancel();
+                Assert.True(twa.IsCompleted);
+                Assert.True(twa.IsCanceled);
+            }
+
+            // Generic, second completes first
+            {
+                var t1 = new TaskCompletionSource<int>();
+                var t2 = new TaskCompletionSource<int>();
+
+                var cts = new CancellationTokenSource();
+                cts.Token.Register(() => t2.TrySetCanceled());
+                cts.Token.Register(() => t1.TrySetCanceled());
+
+                Task<int[]> twa = Task.WhenAll(t1.Task, t2.Task);
+                Assert.False(twa.IsCompleted);
+                cts.Cancel();
+                Assert.True(twa.IsCompleted);
+                Assert.True(twa.IsCanceled);
+            }
+        }
+
+        [Fact]
+        public static void Task_WhenAll_TwoTasks_WakesOnBothCompletionWithExceptionAndCancellation()
+        {
+            Exception e1 = new Exception();
+            CancellationToken ct2 = new CancellationToken(true);
+
+            // Non-generic, first completes first
+            {
+                var t1 = new TaskCompletionSource<int>();
+                var t2 = new TaskCompletionSource<int>();
+
+                Task twa = Task.WhenAll((Task)t1.Task, (Task)t2.Task);
+                Assert.False(twa.IsCompleted);
+                t1.SetException(e1);
+                Assert.False(twa.IsCompleted);
+                t2.SetCanceled(ct2);
+                Assert.True(twa.IsCompleted);
+                Assert.True(twa.IsFaulted);
+                Assert.Equal(e1, twa.Exception?.InnerException);
+                Assert.False(twa.IsCanceled);
```

Comment:
Some of these checks aren't really tests of WhenAll; a Task can't be in multiple states at the same time, so it's a general constraint of Task that IsCanceled would never be true if IsFaulted is true.

Reply:
Useless assertions are removed.